logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java

```diff
@@ -599,9 +609,14 @@ public synchronized Batch nonBlockReadBatch(int limit) throws IOException {
      * @param limit size limit of the batch to read. returned {@link Batch} can be smaller.
      * @param timeout the maximum time to wait in milliseconds on write operations
      * @return the read {@link Batch} or null if no element upon timeout
+     * @throws QueueRuntimeException if queue is closed
      * @throws IOException if an IO error occurs
      */
     public synchronized Batch readBatch(int limit, long timeout) throws IOException {
+        if (isClosed()) {
+            throw new QueueRuntimeException(QueueExceptionMessages.WHILE_READING);
+        }
+
         lock.lock();
         try {
```

Comment:
Both this `readBatch` and the companion `nonBlockReadBatch` use shared and already-guarded `nextReadPage`, so I don't think this pre-check adds value.

```suggestion
        lock.lock();
        try {
```